File: app/src/main/java/com/heinika/designpattern/abstractfactorypattern/FactoryProducer.java
package com.heinika.designpattern.abstractfactorypattern;

import java.lang.reflect.InvocationTargetException;

public class FactoryProducer {
    public AbstractFactory getFactory(String factoryClassName){
        try {
            String FACTORY_PACKAGE = "com.heinika.designpattern.abstractfactorypattern.";
            Class c = Class.forName(FACTORY_PACKAGE +  factoryClassName);
            return (AbstractFactory) c.getConstructor().newInstance();
        } catch (ClassNotFoundException e) {
            e.printStackTrace();
        } catch (InstantiationException e) {
            e.printStackTrace();
        } catch (InvocationTargetException e) {
            e.printStackTrace();
        } catch (NoSuchMethodException e) {
            e.printStackTrace();
        } catch (IllegalAccessException e) {
            e.printStackTrace();
        }
        return null;
    }
}

File: app/src/main/java/com/heinika/designpattern/kotlin/abstractfactory/FactoryProducer.kt
package com.heinika.designpattern.kotlin.abstractfactory

import java.lang.Exception

class FactoryProducer {
    fun <T : AbstractFactory> getFactory(c: Class<T>): AbstractFactory? {
        try {
            return c.getDeclaredConstructor().newInstance()
        } catch (e: Exception) {
            return null
        }
    }
}